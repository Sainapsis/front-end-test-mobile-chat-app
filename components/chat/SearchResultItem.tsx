import { formatMessageDate } from '@/utils/dateUtils';
import { messageSearchStyles } from '@/styles/components/messageSearch.styles';
import { ThemedText } from '@/components/common/ThemedText';
import { View, Pressable } from 'react-native';
import React from 'react';
import { type SearchResult } from '@/hooks/useMessageSearch';

interface SearchResultItemProps {
    item: SearchResult;
    onPress: (chatId: string) => void;
    theme: any;
}

const SearchResultItem = React.memo(({ item, onPress, theme }: SearchResultItemProps) => (
    <Pressable
        style={[messageSearchStyles.resultItem, { borderBottomColor: theme.colors.border }]}
        onPress={() => onPress(item.chatId)}
        accessibilityRole="button"
        accessibilityLabel={`Message from ${item.chatName} at ${formatMessageDate(item.timestamp)}`}
    >
        <View style={messageSearchStyles.resultContent}>
            <View style={messageSearchStyles.messageInfo}>
                <ThemedText style={messageSearchStyles.chatName}>
                    {item.chatName}
                </ThemedText>
                <ThemedText style={[messageSearchStyles.timestamp, { color: theme.colors.textSecondary }]}>
                    {formatMessageDate(item.timestamp)}
                </ThemedText>
            </View>
            <ThemedText
                style={[messageSearchStyles.messageText, { color: theme.colors.textSecondary }]}
                numberOfLines={2}
            >
                {item.text}
            </ThemedText>
        </View>
    </Pressable>
));

export default SearchResultItem;
