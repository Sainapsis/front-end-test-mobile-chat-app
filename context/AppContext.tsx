import React, { createContext, useContext, ReactNode } from 'react';
import { useUser } from '@/hooks/useUser';
import { useChats } from '@/hooks/useChats';
import { DatabaseProvider } from '../database/DatabaseProvider';
import { useDatabase } from '@/hooks/useDatabase';
import { User } from '@/types/User';
import { Chat } from '@/types/Chat';

/**
 * Interface representing the application context state
 */
type AppContextType = {
  users: User[];
  currentUser: User | null;
  isLoggedIn: boolean;
  login: (userId: string) => Promise<boolean>;
  logout: () => void;
  /** List of chats for the current user */
  chats: Chat[];
  createChat: (participantIds: string[]) => Promise<Chat | null>;
  sendMessage: (chatId: string, text: string, senderId: string, imageUri?: string) => Promise<boolean>;
  loading: boolean;
  dbInitialized: boolean;
  clearChats: (userId: string) => Promise<void>;
  deleteChat?: (chatId: string, userId: string) => Promise<void>;
  deleteMessage?: (messageId: string, chatId: string) => Promise<boolean>;
  addReaction?: (messageId: string, emoji: string) => Promise<boolean>;
  removeReaction?: (reactionId: string, messageId: string) => Promise<boolean>;
  editMessage?: (messageId: string, newText: string) => Promise<boolean>;
};

/**
 * Context for managing application-wide state
 */
const AppContext = createContext<AppContextType | undefined>(undefined);

/**
 * Internal component that wraps the application content with context
 * @param children - Child components to be wrapped by the context
 */
function AppContent({ children }: { children: ReactNode }) {
  const { isInitialized } = useDatabase();
  const userContext = useUser();
  const chatContext = useChats(userContext.currentUser?.id || null);
  
  const loading = !isInitialized || userContext.loading || chatContext.loading;

  const value: AppContextType = {
    ...userContext,
    ...chatContext,
    loading,
    dbInitialized: isInitialized,
    deleteChat: chatContext.deleteChat || undefined,
    deleteMessage: chatContext.deleteMessage || undefined,
    addReaction: chatContext.addReaction || undefined,
    removeReaction: chatContext.removeReaction || undefined,
    editMessage: chatContext.editMessage || undefined,
    sendMessage: async (chatId: string, text: string, senderId: string, imageUri?: string) => {
      if (!chatContext.sendMessage) return false;
      return chatContext.sendMessage(chatId, text, senderId, imageUri);
    },
  };

  if (!isInitialized) {
    return null;
  }

  return <AppContext.Provider value={value}>{children}</AppContext.Provider>;
}

/**
 * Provider component that wraps the application with necessary contexts
 * @param children - Child components to be wrapped by the provider
 */
export const AppProvider = ({ children }: { children: React.ReactNode }) => (
  <DatabaseProvider>
    <AppContent>{children}</AppContent>
  </DatabaseProvider>
);

/**
 * Hook for accessing application context
 * @returns Application context containing user, chat, and database states
 * @throws Error if used outside of AppProvider
 */
export function useAppContext() {
  const context = useContext(AppContext);
  if (!context) {
    throw new Error('useAppContext must be used within an AppProvider');
  }
  return context;
}
