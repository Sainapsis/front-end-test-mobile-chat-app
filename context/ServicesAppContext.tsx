import React, { createContext, useContext, ReactNode } from 'react';
import { DatabaseProvider, useDatabaseStatus } from './DataBaseContext';
import { useChatsDb,  useUserDb } from '../hooks/db';
import { Chat, Message, User } from '@/utils/types';

type AppContextType = {
  users: User[];
  currentUser: User | null;
  isLoggedIn: boolean;
  login: (userId: string) => Promise<boolean>;
  logout: () => void;
  chats: Chat[];
  createChat: (participantIds: string[]) => Promise<Chat | null>;
  sendMessage: (
    chatId: string, 
    text: string | undefined, 
    senderId: string,
    mediaUrl?: string,
    mediaType?: string,
    mediaThumbnail?: string,
    voiceUrl?: string,
    voiceDuration?: number,
    isVoiceMessage?: boolean,
    
  ) => Promise<Message | null>;
  addReaction: (messageId: string, userId: string, emoji: string) => Promise<boolean>;
  removeReaction: (reactionId: string) => Promise<boolean>;
  searchMessages: (query: string, chatId?: string) => Promise<void>;
  searchResults: Message[];
  deleteMessage: (messageId: string) => Promise<boolean>;
  editMessage: (messageId: string, newText: string) => Promise<boolean>;
  updateMessageStatus: (messageId: string, status: 'sent' | 'delivered' | 'read') => Promise<void>;
  markMessagesAsRead: (chatId: string) => Promise<void>;
  loading: boolean;
  dbInitialized: boolean;
};

export const AppContext = createContext<AppContextType | undefined>(undefined);

function AppContent({ children }: { children: ReactNode }) {
  const { isInitialized } = useDatabaseStatus();
  const userContext = useUserDb();
  const chatContext = useChatsDb(userContext.currentUser?.id || null);
  
  const loading = !isInitialized || userContext.loading || chatContext.loading;

  const value = {
    ...userContext,
    ...chatContext,
    loading,
    dbInitialized: isInitialized,
  };

  return <AppContext.Provider value={value}>{children}</AppContext.Provider>;
}

export function AppProvider({ children }: { children: ReactNode }) {
  return (
    <DatabaseProvider>
      <AppContent>{children}</AppContent>
    </DatabaseProvider>
  );
}

