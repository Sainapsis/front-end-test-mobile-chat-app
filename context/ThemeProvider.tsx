import React, { useState, useEffect } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';
import ThemeContext from './ThemeContext';
import { Theme } from '@/types/tColores';

/**
 * Provider component that handles theme state and persistence
 * @param children - Child components to be wrapped by the provider
 */
export const ThemeProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const [theme, setTheme] = useState<Theme>('light');

  useEffect(() => {
    loadTheme();
  }, []);

  const loadTheme = async () => {
    try {
      const savedTheme = await AsyncStorage.getItem('theme');
      if (savedTheme) {
        setTheme(savedTheme as Theme);
      }
    } catch (error) {
      console.error('Error loading theme:', error);
    }
  };

  const toggleTheme = async () => {
    const newTheme = theme === 'light' ? 'dark' : 'light';
    setTheme(newTheme);
    try {
      await AsyncStorage.setItem('theme', newTheme);
    } catch (error) {
      console.error('Error saving theme:', error);
    }
  };

  return (
    <ThemeContext.Provider value={{ theme, toggleTheme }}>
      {children}
    </ThemeContext.Provider>
  );
};