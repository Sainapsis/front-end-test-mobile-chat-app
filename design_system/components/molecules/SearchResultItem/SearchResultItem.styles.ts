import { StyleSheet } from 'react-native';
import { colors, spacing } from '@/design_system/ui/tokens';

export const styles = StyleSheet.create({
  resultItem: {
    marginBottom: spacing.md,
    padding: spacing.sm,
    borderRadius: 8,
    borderWidth: 1,
    borderColor: colors.neutral[200],
    shadowColor: colors.neutral[900],
    shadowOffset: {
      width: 0,
      height: 1,
    },
    shadowOpacity: 0.1,
    shadowRadius: 2,
    elevation: 2,
  },
  messageHeader: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    marginBottom: spacing.xs,
  },
  headerLeft: {
    flexDirection: 'row',
    alignItems: 'center',
    flex: 1,
  },
  chatName: {
    fontSize: 14,
    fontWeight: '500',
    marginLeft: spacing.xs,
    flex: 1,
  },
  timestamp: {
    fontSize: 12,
    marginLeft: spacing.sm,
  },
});