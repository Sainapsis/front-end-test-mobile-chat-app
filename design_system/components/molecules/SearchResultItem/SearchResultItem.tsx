import React from 'react';
import { View, TouchableOpacity } from 'react-native';
import { ThemedText } from '@/design_system/components/atoms/ThemedText';
import { MessageBubble } from '@/design_system/components/organisms/MessageBubble';
import { IconSymbol } from '@/design_system/ui/vendors/IconSymbol';
import { colors, themes as Colors } from '@/design_system/ui/tokens';
import { styles } from './SearchResultItem.styles';
import { useColorScheme } from '@/hooks/useColorScheme';

interface SearchResultItemProps {
  /** Search result item data */
  item: any;
  /** Current user's name */
  currentUserName?: string;
  /** Current user's ID */
  currentUserId: string;
  /** Function to be called when the item is pressed */
  onPress: () => void;
}

/**
 * SearchResultItem component displays a single search result in a list.
 * It shows the chat name, timestamp, and a preview of the message.
 */
export function SearchResultItem ({
  item,
  currentUserName,
  currentUserId,
  onPress,
}:SearchResultItemProps) {
  const theme = useColorScheme() ?? 'light';

  return (
    <TouchableOpacity style={[theme === 'light' ? {backgroundColor: Colors.light.text.secondary} : {backgroundColor:Colors.dark.text.secondary},styles.resultItem]} onPress={onPress}>
    <View style={styles.messageHeader}>
      <View style={styles.headerLeft}>
        <IconSymbol 
          name="chevron.right" 
          size={16} 
          color={colors.neutral[400]} 
        />
        <ThemedText style={[theme === 'light' ? {color: Colors.light.text.contrast} : {color:Colors.dark.text.contrast},styles.chatName]}>
          Chat with {item.participant_names?.split(',')
            .find((name: string) => name !== currentUserName)?.trim()}
        </ThemedText>
      </View>
      <ThemedText style={[theme === 'light' ? {color: Colors.light.text.secondary} : {color:Colors.dark.text.secondary},styles.timestamp]}>
        {new Date(item.timestamp).toLocaleTimeString([], { 
          hour: '2-digit', 
          minute: '2-digit' 
        })}
      </ThemedText>
    </View>
    <MessageBubble
      message={item}
      isCurrentUser={item.sender_id === currentUserId}
      userId={currentUserId}
    />
  </TouchableOpacity>
  );
}