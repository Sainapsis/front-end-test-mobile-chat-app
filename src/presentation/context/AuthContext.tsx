import { LoginParams } from "@/src/data/interfaces/auth.interface";
import { createContext, ReactNode, useContext } from "react";
import { useAuth } from "../hooks/useAuth";
import { User } from "@/src/domain/entities/user";

interface AuthContextType {
  currentUser: User | null;
  login: ({ userId }: LoginParams) => Promise<boolean>;
  logout: () => void;
}

const _AuthContext = createContext<AuthContextType | undefined>(undefined);

export function AuthContext({ children }: { children: ReactNode }) {
  const authContext = useAuth();

  const value = {
    ...authContext,
  };

  return (
    <_AuthContext.Provider value={value}>{children}</_AuthContext.Provider>
  );
}

export function useAuthContext() {
  const context = useContext(_AuthContext);

  if (context === undefined) {
    throw new Error("useAuthContext must be used within an AppProvider");
  }

  return context;
}
