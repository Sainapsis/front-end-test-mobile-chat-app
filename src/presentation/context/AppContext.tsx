import React, { createContext, useContext, ReactNode } from "react";
import { useDatabase } from "../hooks/useDatabase";
import { useUser } from "@/src/presentation/hooks/useUser";
import { DatabaseContext } from "@/src/presentation/context/DatabaseProvider";
import { UserContext } from "./UserContext";
import { AuthContext } from "./AuthContext";

interface AppContextType {
  dbInitialized: boolean;
}

const AppContext = createContext<AppContextType | undefined>(undefined);

function AppContent({ children }: { children: ReactNode }) {
  const { isInitialized } = useDatabase();
  const userContext = useUser();
  
  const value = {
    ...userContext,
    dbInitialized: isInitialized,
  };

  return <AppContext.Provider value={value}>{children}</AppContext.Provider>;
}

export function AppProvider({ chatId, children }: { chatId: string; children: ReactNode }) {
  return (
    <DatabaseContext>
      <AuthContext>
        <UserContext>
          <AppContent>{children}</AppContent>
        </UserContext>
      </AuthContext>
    </DatabaseContext>
  );
}

export function useAppContext() {
  const context = useContext(AppContext);

  if (context === undefined) {
    throw new Error("useAppContext must be used within an AppProvider");
  }

  return context;
}
