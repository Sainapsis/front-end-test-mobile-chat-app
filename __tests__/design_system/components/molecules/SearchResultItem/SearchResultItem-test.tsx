import * as React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import { ThemeProvider } from '@/context/ThemeContext';
import { SearchResultItem } from '@/design_system/components/molecules/SearchResultItem/SearchResultItem';
// Mock para MessageBubble

jest.mock('@/design_system/components/organisms/MessageBubble', () => {
    const { View, Text } = require('react-native');
  
    return {
      MessageBubble: ({ message }: { message: { content: string } }) => (
        <View testID="message-bubble">
          <Text>{message.content}</Text>
        </View>
      )
    };
  });
  


// Mock para IconSymbol
jest.mock('@/design_system/ui/vendors/IconSymbol', () => {
    const { View } = require('react-native');
  
    return {
      IconSymbol: ({ name }: { name: string }) => (
        <View testID={`icon-${name}`} />
      ),
    };
  });
  

// Mock para useColorScheme
jest.mock('@/hooks/useColorScheme', () => ({
  useColorScheme: jest.fn(() => 'light')
}));

describe('SearchResultItem Component', () => {
  const mockItem = {
    id: '123',
    participant_names: 'John Doe,Jane Smith',
    sender_id: 'user2',
    timestamp: '2023-05-15T14:30:00Z',
    content: 'Hello, this is a test message'
  };

  const defaultProps = {
    item: mockItem,
    currentUserName: 'Jane Smith',
    currentUserId: 'user1',
    onPress: jest.fn()
  };

  const renderWithTheme = (component: string | number | boolean | Iterable<React.ReactNode> | React.JSX.Element | null | undefined) => {
    return render(
      <ThemeProvider>
        {component}
      </ThemeProvider>
    );
  };

  it('renders correctly with snapshot', () => {
    const tree = renderer.create(
      <ThemeProvider>
        <SearchResultItem {...defaultProps} />
      </ThemeProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('displays the correct chat name', () => {
    const { getByText } = renderWithTheme(
      <SearchResultItem {...defaultProps} />
    );
    
    // Debería mostrar el nombre del otro participante
    expect(getByText(/Chat with John Doe/)).toBeTruthy();
  });


  it('renders MessageBubble with correct props', () => {
    const { getByTestId } = renderWithTheme(
      <SearchResultItem {...defaultProps} />
    );
    
    expect(getByTestId('message-bubble')).toBeTruthy();
  });

  it('renders with correct icon', () => {
    const { getByTestId } = renderWithTheme(
      <SearchResultItem {...defaultProps} />
    );
    
    expect(getByTestId('icon-chevron.right')).toBeTruthy();
  });

  it('displays the correct timestamp', () => {
    // Crear un mock de Date.prototype.toLocaleTimeString
    const originalToLocaleTimeString = Date.prototype.toLocaleTimeString;
    Date.prototype.toLocaleTimeString = jest.fn(() => '2:30 PM');
    
    const { getByText } = renderWithTheme(
      <SearchResultItem {...defaultProps} />
    );
    
    expect(getByText('2:30 PM')).toBeTruthy();
    
    // Restaurar el método original
    Date.prototype.toLocaleTimeString = originalToLocaleTimeString;
  });
});

function queryByTestId(arg0: string) {
    throw new Error('Function not implemented.');
}
