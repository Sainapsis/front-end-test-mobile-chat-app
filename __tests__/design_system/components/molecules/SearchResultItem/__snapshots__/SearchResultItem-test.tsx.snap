// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResultItem Component renders correctly with snapshot 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "#687076",
      "borderColor": "#EEEEEE",
      "borderRadius": 8,
      "borderWidth": 1,
      "elevation": 2,
      "marginBottom": 12,
      "opacity": 1,
      "padding": 8,
      "shadowColor": "#212121",
      "shadowOffset": {
        "height": 1,
        "width": 0,
      },
      "shadowOpacity": 0.1,
      "shadowRadius": 2,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 4,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
        }
      }
    >
      <View
        testID="icon-chevron.right"
      />
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "fontWeight": "400",
              "lineHeight": 24,
            },
            [
              {
                "color": "#FFFFFF",
              },
              {
                "flex": 1,
                "fontSize": 14,
                "fontWeight": "500",
                "marginLeft": 4,
              },
            ],
          ]
        }
      >
        Chat with 
        John Doe
      </Text>
    </View>
    <Text
      style={
        [
          {
            "color": "#11181C",
          },
          {
            "fontSize": 16,
            "fontWeight": "400",
            "lineHeight": 24,
          },
          [
            {
              "color": "#687076",
            },
            {
              "fontSize": 12,
              "marginLeft": 8,
            },
          ],
        ]
      }
    >
      09:30 AM
    </Text>
  </View>
  <View
    testID="message-bubble"
  >
    <Text>
      Hello, this is a test message
    </Text>
  </View>
</View>
`;
